druid-handler/src/java/org/apache/hadoop/hive/ql/io/DruidOutputFormat.java

```diff
@@ -0,0 +1,393 @@
+                                             public String getFormatString()
+                                             {
+                                               return "processing-%s";
+                                             }
+
+                                             @Override
+                                             public int intermediateComputeSizeBytes()
+                                             {
+                                               return 100 * 1024 * 1024;
+                                             }
+
+                                             @Override
+                                             public int getNumThreads()
+                                             {
+                                               return 1;
+                                             }
+
+                                             @Override
+                                             public int columnCacheSizeBytes()
+                                             {
+                                               return 25 * 1024 * 1024;
+                                             }
+                                           }
+                                       );
+                                       binder.bind(ColumnConfig.class).to(DruidProcessingConfig.class);
+                                     }
+                                   }
+          )
+      );
+      ObjectMapper objectMapper = injector.getInstance(ObjectMapper.class);
+      AppenderatorFactory defaultOfflineAppenderatorFactory = null;
+      try {
+        defaultOfflineAppenderatorFactory = objectMapper.readerFor(AppenderatorFactory.class)
+                                                        .readValue("{\"type\":\"offline\"}");
+      }
+      catch (IOException e) {
+        Throwables.propagate(e);
+      }
+      this.tuningConfig = Preconditions.checkNotNull(realtimeTuningConfig);
+
+
+      this.dataSchema = dataSchema;
+      appenderator = defaultOfflineAppenderatorFactory.build(
+          this.dataSchema,
+          tuningConfig,
+          new FireDepartmentMetrics()
+      );
+      this.maxPartitionSize = maxPartitionSize == null ? DEFAULT_MAX_PARTITION_SIZE : maxPartitionSize;
+
+      appenderator.startJob(); // maybe we need to move this out of the constructor
+    }
+
+    private SegmentIdentifier getSegmentIdentifier(DruidWritable druidHiveRecord)
+    {
+      return getSegmentIdentifier(Longs.tryParse((String) druidHiveRecord.getValue()
+                                                                         .get(DruidTable.DEFAULT_TIMESTAMP_COLUMN)));
+    }
+
+    private SegmentIdentifier getSegmentIdentifier(long timestamp)
+    {
+
+      final Granularity segmentGranularity = dataSchema.getGranularitySpec().getSegmentGranularity();
+
+
+      final long truncatedTime = segmentGranularity.truncate(new DateTime(timestamp)).getMillis();
```

Comment:
Since we have to distribute (partition) the records depending on segment granularity, I think it is not necessary to compute here the truncated time. Rather, we could point to the column containing the truncated timestamp in the record (which could be simply the last field in the record). That way, we avoid computation for each record that we are writing, which can be expensive.